### ISSUE TITLE
SQL:2023 support matrix for ClickHouse

### ISSUE THREAD

**Issue by UnamedRus** [ClickHouse/ClickHouse#48388]
Based on http://peter.eisentraut.org/blog/2023/04/04/sql-2023-is-finished-here-is-whats-new review (kudos for him)

ClickHouse 25.10

### UNIQUE null treatment (F292) (Not supported)

Unique constants does not exist in ClickHouse

### ORDER BY in grouped table (F868) (Not supported)

```
SELECT
    product.product_id,
    sum(product_part.num)
FROM
(
    SELECT
        1 AS product_id,
        'aaa' AS product_code
) AS product
INNER JOIN
(
    SELECT
        1 AS product_id,
        10 AS num
) AS product_part ON product.product_id = product_part.product_id
GROUP BY product.product_id
ORDER BY product.product_code ASC

Code: 215. DB::Exception: Received from localhost:9000. DB::Exception: Column 'product.product_code' is not under aggregate function and not in GROUP BY keys. In query SELECT product.product_id, sum(product_part.num) FROM (SELECT 1 AS product_id, 'aaa' AS product_code) AS product INNER JOIN (SELECT 1 AS product_id, 10 AS num) AS product_part ON product.product_id = product_part.product_id GROUP BY product.product_id ORDER BY product.product_code ASC. (NOT_AN_AGGREGATE)
```

### GREATEST and LEAST (T054) (Supported)

```
SELECT GREATEST(1, 2, 3)

┌─greatest(1, 2, 3)─┐
│                 3 │
└───────────────────┘
```

### String padding functions (T055) (Supported)

```
SELECT lpad(CAST('12345.67', 'varchar'), 12, '*') AS res

Query id: 1ee3a790-8edc-4ce5-9db4-92f5b26b40b3

┌─res──────────┐
│ ****12345.67 │
└──────────────┘
```

### Multi-character TRIM functions (T056) (Partial)

Supported, but with old and "heavy" syntax

```
SELECT TRIM(LEADING '*' FROM '****12345.67') as res;

┌─res──────┐
│ 12345.67 │
└──────────┘

SELECT ltrim('*','****12345.67') as res;

Expected one of:
```

Need alias for ltrim function with 2 arguments

https://github.com/ClickHouse/ClickHouse/issues/90170
https://github.com/ClickHouse/ClickHouse/pull/90305

### Optional string types maximum length (T081) (Supported)

```
CREATE TABLE t1
(
    `a` VARCHAR,
    `b` VARCHAR
)
ENGINE = Memory
```

### Enhanced cycle mark values (T133) (Not supported)

Recursive CTE's are supported

CYCLE mark is not supported

https://github.com/ClickHouse/ClickHouse/blob/dbb20fc864216f4dad09b20330619fbd736c7fd2/src/Analyzer/RecursiveCTE.h#L34

### ANY_VALUE (T626) (Supported)

```
SELECT any_value(number)
FROM numbers(10)

Query id: 39579c85-2850-4696-8d4f-7f04bcf73ad3

   ┌─any_value(number)─┐
1. │                 0 │
   └───────────────────┘
```

(any function exist in ClickHouse)
https://github.com/ClickHouse/ClickHouse/pull/52147
https://github.com/ClickHouse/ClickHouse/issues/52140

### Non-decimal integer literals (T661) (Partial)

Working:

```
SELECT 0xFFFF,  0b11001111;
```

Non-working: 

```
SELECT 0o755

Received exception:
Code: 47. DB::Exception: Missing columns: '0o755'
```

### Underscores in numeric literals (T662) (Partial)

Working:

```
SELECT 1_000_000;
```

Non working:

```
SELECT
    1000000,
    `0x_FFFF_FFFF`

Code: 47. DB::Exception: Missing columns: '0x_FFFF_FFFF'
```

### JSON data type (T801) (Supported)

Experimental implementation of Object('JSON') was deprecated and replaced by new implementation
https://github.com/ClickHouse/ClickHouse/pull/66444

### Enhanced JSON data type (T802) (Partial)

Allow to add constrain on only uniq keys 
JSON('...text...' WITH UNIQUE KEYS).

(BTW, it's default behavior for ClickHouse)

```
SELECT CAST('{"foo": {"bar": [100, 200, 300]}, "foo": "e" }', 'JSON') AS a


Elapsed: 0.005 sec.

Received exception from server (version 25.10.1):
Code: 117. DB::Exception: Received from localhost:9000. DB::Exception: Cannot insert data into JSON column: Duplicate path found during parsing JSON object: foo. You can enable setting type_json_skip_duplicated_paths to skip duplicated paths during insert: In scope SELECT CAST('{"foo": {"bar": [100, 200, 300]}, "foo": "e" }', 'JSON') AS a. (INCORRECT_DATA)
```

### String-based JSON (T803) (Not supported)

Need an alias data type

### Hex integer literals in SQL/JSON path language (T840) (Not supported)

???

### SQL/JSON simplified accessor (T860–T864) (Supported)

```
SELECT k.foo.bar[2]
FROM json

┌─arrayElement(k.foo.bar, 2)─┐
│                        200 │
└────────────────────────────┘

SELECT
    CAST('{"foo": {"bar": [100, 200, 300]}, "aaa": "e" }', 'JSON') AS a,
    a.foo.bar[2]

Query id: 205c5eeb-c843-48d6-b3aa-1253c761d2ac

   ┌─a───────────────────────────────────────┬─arrayElement(a.foo.bar, 2)─┐
1. │ {"aaa":"e","foo":{"bar":[100,200,300]}} │ 200                        │
   └─────────────────────────────────────────┴────────────────────────────┘
```

### SQL/JSON item methods (T865–T878) (Not supported)

```
    T865: SQL/JSON item method: bigint()
    T866: SQL/JSON item method: boolean()
    T867: SQL/JSON item method: date()
    T868: SQL/JSON item method: decimal()
    T869: SQL/JSON item method: decimal() with precision and scale
    T870: SQL/JSON item method: integer()
    T871: SQL/JSON item method: number()
    T872: SQL/JSON item method: string()
    T873: SQL/JSON item method: time()
    T874: SQL/JSON item method: time_tz()
    T875: SQL/JSON item method: time precision
    T876: SQL/JSON item method: timestamp()
    T877: SQL/JSON item method: timestamp_tz()
    T878: SQL/JSON item method: timestamp precision
```

### JSON comparison (T879–T882) (Supported)

```
SELECT
    CAST('{"foo": {"bar": [100, 200, 300]}, "aaa": "e" }', 'JSON') AS a,
    CAST('{"foo": "result"}', 'JSON') AS b,
    CAST('{"foo": "result"}', 'JSON') AS c,
    a = b,
    b = c

   ┌─a───────────────────────────────────────┬─b────────────────┬─c────────────────┬─equals(a, b)─┬─equals(b, c)─┐
1. │ {"aaa":"e","foo":{"bar":[100,200,300]}} │ {"foo":"result"} │ {"foo":"result"} │            0 │            1 │
   └─────────────────────────────────────────┴──────────────────┴──────────────────┴──────────────┴──────────────┘
```

## Property Graph Queries (SQL/PGQ) (Not Supported)

```
CREATE PROPERTY GRAPH financial_transactions
    VERTEX TABLES (person, company, account)
    EDGE TABLES (ownerof, transaction);

SELECT owner_name,
       SUM(amount) AS total_transacted
FROM financial_transactions GRAPH_TABLE (
  MATCH (p:person WHERE p.name = 'Alice')
        -[:ownerof]-> (:account)
        -[t:transaction]- (:account)
        <-[:ownerof]- (owner:person|company)
  COLUMNS (owner.name AS owner_name, t_amount AS amount)
) AS ft
GROUP BY owner_name;
```

Not supported

---

Should be easy to implement / fix:

* Multi-character TRIM functions (T056) (Partial)
* Non-decimal integer literals (T661) (Partial)
* Underscores in numeric literals (T662) (Partial)

**Comment by mrk-andreev** [comment]
# Graph Queries in ClickHouse via SQL/PGQ

## TL;DR

Two-phase proposal to bring graph queries to ClickHouse as a SQL extension — no separate query language, no new storage engine, no migration cost.

**Phase 1** — query a graph inline using a `WITH GRAPH` statement scoped to the query. Everything lives in one statement, nothing persists.

**Phase 2** — persistent `CREATE GRAPH` registered in the ClickHouse catalog, like dictionaries. Survives restarts, visible to all connections.

---

## Background

We looked at what graph capabilities ClickHouse lacks compared to Neo4j, Oracle, and TigerGraph, and how other engines approached the problem.

Oracle's `CONNECT BY` was the natural starting point — it covers trees and simple hierarchical traversal well. But it breaks the moment a node has multiple parents (fan-in): it emits one row per path rather than one row per node, so results duplicate in diamond-shaped graphs. It is a good inspiration, not a solution.

Oracle Database 23ai, DuckDB (via DuckPGQ), and BigQuery have each adopted graph query capabilities based on the same underlying standard: **ISO SQL:2023 Part 16 (SQL/PGQ)** — the only graph query standard with ISO backing. SQL Server has a separate, proprietary graph extension using `MATCH` in the `WHERE` clause, predating and architecturally distinct from SQL/PGQ. The gap in ClickHouse is known; there is just no proposal yet.

The closest architectural reference for ClickHouse is **DuckPGQ** — a community extension implementing SQL/PGQ on top of DuckDB's vectorised columnar engine, which is exactly what ClickHouse is. Their CIDR 2023 paper documents the full design.

---

## The Proposal

### Phase 1 — query-scoped graph, discarded after use

```sql
WITH GRAPH my_graph AS (
    VERTEX persons AS
    (
        TABLE persons
        KEY id
        COLUMNS (name)
    ),

    EDGE transfers AS
    (
        TABLE transfers
        KEY id
        FROM from_id REFERENCES persons.id
        TO   to_id   REFERENCES persons.id
        COLUMNS (amount)
    )

    )
SELECT p1.name, p2.name, t.amount
FROM GRAPH_TABLE (my_graph
    MATCH (p1:persons)-[t:transfers]->(p2:persons)
    COLUMNS (p1.name, p2.name, t.amount)
)
WHERE t.amount > 10000;
```

The graph schema is defined with `CREATE PROPERTY GRAPH`, consistent with the ISO SQL/PGQ standard. `GRAPH_TABLE` returns a normal relation — it composes with `WHERE`, `GROUP BY`, `JOIN`, and window functions as usual. In Phase 1, the graph is session-scoped and not persisted to the catalog. No schema side effects survive the session.

### Phase 2 — persistent named graph (global, like dictionaries)

```sql
CREATE GRAPH VIEW bank_graph AS (
    VERTEX persons AS
    (
        TABLE persons
        KEY id
        COLUMNS (name)
    ),

    EDGE transfers AS
    (
        TABLE transfers
        KEY id
        FROM from_id REFERENCES persons.id
        TO   to_id   REFERENCES persons.id
        COLUMNS (amount)
    )
);

SELECT owner, SUM(amount)
FROM GRAPH_TABLE (bank_graph
    MATCH (p:persons)-[t:transfers*1..3]->(a:accounts)
    COLUMNS (p.name AS owner, t.amount AS amount)
)
GROUP BY owner;
```

`CREATE GRAPH` registers a named graph schema in the ClickHouse catalog, the same way `CREATE DICTIONARY` works. It is purely metadata — no data is copied, no new storage, just a mapping over existing tables. Survives restarts, visible to all connections, droppable with `DROP GRAPH`.

---

## Why it fits ClickHouse

- Works over existing MergeTree tables — nothing to migrate, nothing to remodel
- Phase 1 is purely a parser + query rewriter change, no storage or execution engine modifications
- Phase 2 follows the dictionary mental model already familiar to ClickHouse users
- ISO standard means it is not a bet on proprietary syntax that might go nowhere
- DuckPGQ already validated this architecture on a columnar vectorised engine — the mapping to ClickHouse is direct

---

Does this direction look reasonable? Would the core team accept a community PR for Phase 1, and are there parser or planner constraints we should be aware of before starting?
